confluent_kafka/schema_registry/protobuf.py

```diff
@@ -0,0 +1,437 @@
+            schema_refs.append(SchemaReference(dep.name,
+                                               subject,
+                                               reference.version))
+        return schema_refs
+
+    def __call__(self, msg, ctx):
+        """
+        Serializes a Protobuf Message to the Confluent Schema Registry
+        Protobuf binary format.
+
+        Args:
+            msg (Message): Protobuf message instance.
+
+            ctx (SerializationContext): Metadata pertaining to the serialization
+                operation.
+
+        Note:
+            None objects are represented as Kafka Null.
+
+        Raises:
+            SerializerError if any error occurs serializing obj
+
+        Returns:
+            bytes: Confluent Schema Registry formatted Protobuf bytes
+
+        """
+        if msg is None:
+            return None
+
+        subject = self._subject_name_func(ctx, msg.DESCRIPTOR.full_name)
+
+        if subject not in self._known_subjects:
+            self._schema.references = self.resolve_dependencies(ctx, msg.DESCRIPTOR.file)
+
+        self._schema_id = self._registry.register_schema(subject, self._schema)
+
+        with _ContextStringIO() as fo:
+            # Write the magic byte and schema ID in network byte order (big endian)
+            fo.write(struct.pack('>bI', _MAGIC_BYTE, self._schema_id))
+            # write the record index to the buffer
+            self._encode_index(fo, self.msg_index)
+            # write the record itself
+            fo.write(msg.SerializeToString())
+            return fo.getvalue()
+
+
+class ProtobufDeserializer(object):
+    """
+    ProtobufDeserializer decodes bytes written in the Schema Registry
+    Protobuf format to an object.
+
+    Args:
+        descriptor (Descriptor): Protobuf Message Descriptor.
+
+    .. _Schema definition:
+        https://json-schema.org/understanding-json-schema/reference/generic.html
+
+    """
+    def __init__(self, descriptor):
+        self.msg_index = _create_msg_index(descriptor)
+        self.msg_class = MessageFactory().GetPrototype(descriptor)
+
+    @staticmethod
+    def _decode_uvarint(buf):
+        """
+        Decodes a single uvarint from a buffer.
+
+        Args:
+            - buf (BytesIO): buffer to read from
+
+        Returns:
+            int: decoded uvarint
+
+        Raises:
+            EOFError: if buffer is empty
+
+        """
+        value = 0
+        shift = 0
+        while True:
+            i = ProtobufDeserializer._read_byte(buf)
+
+            value |= (i & 0x7f) << shift
+            shift += 7
+            if not (i & 0x80):
+                break
+
+        return value
+
+    @staticmethod
+    def _read_byte(buf):
+        """
+        Returns int representation for a byte.
+
+        Args:
+            buf (BytesIO): buffer to read from
+
+        .. _ord:
+            https://docs.python.org/2/library/functions.html#ord
+        """
+        i = buf.read(1)
```

Comment:
seems like there should be some better way of doing this (i don't know, just speaking what i'm thinking)

Reply:
bit nit, it'll work...

Reply:
There are a few annoying things to account for here. 

1. BytesIO.read(1) lets you know that you have hit EOF by returning an empty bytes object.
2. You can't actually perform bitwise operations on a bytes object (hence ord)

Reply:
resolving for now we  can address later if we decide on a cleaner suggestion. I don't want to block on cosmetics though.